### PR TITLE
[MXNET-423] Gluon Model Zoo Pre Trained Model tutorial

### PR DESCRIPTION
**Pull request by ThomasDelteil**
## Description ##
There has been several instances of beginner mxnet users who haven't been able to use the models from the model zoo as they were lacking a fully integrated example. This puts all parts together. There was a blog post recently posted here: https://aws.amazon.com/fr/blogs/machine-learning/use-pre-trained-models-with-apache-mxnet/ using the module API. This is an attempt to draw people towards the nicer Gluon API for pre-trained model.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by piiswrong** [comment]
BTW where is the entry point for these tutorials? Are they listed on the tutorial page?

**Comment by ThomasDelteil** [comment]
@piiswrong yes the index.md file of the docs/tutorial folder is updated

File: docs/tutorials/index.md [hunk 1]
```diff
@@ -32,8 +32,8 @@ Select API:&nbsp;
     * [60-Minute Gluon Crash Course](http://gluon-crash-course.mxnet.io/) <img src="https://upload.wikimedia.org/wikipedia/commons/6/6a/External_link_font_awesome.svg" alt="External link" height="15px" style="margin: 0px 0px 3px 3px;"/>
     * [MNIST Handwritten Digit Classification](/tutorials/gluon/mnist.html)
 * Models
+    * [Model Zoo: using pre-trained models](/tutorials/gluon/pretrained_models.html)
     * [Linear Regression](http://gluon.mxnet.io/chapter02_supervised-learning/linear-regression-gluon.html) <img src="https://upload.wikimedia.org/wikipedia/commons/6/6a/External_link_font_awesome.svg" alt="External link" height="15px" style="margin: 0px 0px 3px 3px;"/>
-    * [MNIST Handwritten Digit Classification](/tutorials/gluon/mnist.html)
```
Comment (ThomasDelteil) [inline docs/tutorials/index.md]: this tutorial is already listed three lines above

File: docs/tutorials/gluon/pretrained_models.md [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+import matplotlib.pyplot as plt
+import numpy as np
+import json
+%matplotlib inline
+```
+
+## Loading the model
+
+The [Gluon Model Zoo](https://mxnet.incubator.apache.org/api/python/gluon/model_zoo.html) provides a collection of off-the-shelf models. You can get the ImageNet pre-trained model by using `pretrained=True`. 
+If you want to train on your own classification problem from scratch, you can get an untrained network with a specific number of classes using the `classes=10` for example
```
Comment (indhub) [inline docs/tutorials/gluon/pretrained_models.md]: using the `classes` parameter. For example,

File: docs/tutorials/gluon/pretrained_models.md [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+
+
+Get a test image
+
+
+```python
+filename = mx.test_utils.download('https://github.com/dmlc/web-data/blob/master/mxnet/doc/tutorials/onnx/images/dog.jpg?raw=true', fname='dog.jpg')
+```
+
+If you want to use your own image for the test, upload it in the same folder and change the following line
```
Comment (indhub) [inline docs/tutorials/gluon/pretrained_models.md]: I assume user has to copy the image to the same folder as the notebook. 'Upload' usually means uploading to a server. Do you want to change this to something along the lines of "copy the image to the same folder that contains the notebook and change the following line"?

File: docs/tutorials/gluon/pretrained_models.md [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+```python
+image = mx.image.imread(filename)
+plt.imshow(image.asnumpy())
+```
+
+![png](https://github.com/dmlc/web-data/blob/master/mxnet/doc/tutorials/onnx/images/dog.jpg?raw=true)
+
+
+Neural network expects input in a specific format. Usually images comes in the `Width x Height x Channels` format. Where channels are the RGB channels.
+This network accepts images in the `BatchSize x 3 x 224 x 224`. `224 x 224` is the image resolution, that?s how the model was trained. `3` is the number of channels : Red, Green and Blue (in this order). In this case we use a `BatchSize` of `1` since we are predicting one image at a time.
```
Comment (indhub) [inline docs/tutorials/gluon/pretrained_models.md]: that's

File: docs/tutorials/gluon/pretrained_models.md [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+
+
+We then take the top `k` predictions for our image, here the top `3`.
+
+
+```python
+top_pred = predictions.topk(k=3)[0].asnumpy()
+```
+
+And we print the category predicted with the corresponding probability:
```
Comment (indhub) [inline docs/tutorials/gluon/pretrained_models.md]: categories predicted with their corresponding probabilities.

**Comment by safrooze** [comment]
Regardless of your personal experience, freezing parameters is a technique used in many papers and I'm sure some users would be looking for a way to experiment with that. I think it makes sense to add it to this tutorial.

**Comment by ThomasDelteil** [comment]
@safrooze I am not disputing the fact that freezing parameters is a useful (and used) technique, especially training difficult to converge networks like GANs where you want to freeze parts of your network at times to help convergence.

However I think it is out of scope of a tutorial called **Using pre-trained models from the Gluon Model Zoo**.  I would stick to my decision to not include a fully worked out example of fine-tuning, freezing parameters, etc. It should be the subject of a separate tutorial where its usefulness is demonstrable and demonstrated.

**Comment by safrooze** [comment]
Agreed.

File: docs/tutorials/gluon/pretrained_models.md [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+
+
+Neural network expects input in a specific format. Usually images comes in the `Width x Height x Channels` format. Where channels are the RGB channels.
+This network accepts images in the `BatchSize x 3 x 224 x 224`. `224 x 224` is the image resolution, that's how the model was trained. `3` is the number of channels : Red, Green and Blue (in this order). In this case we use a `BatchSize` of `1` since we are predicting one image at a time.
+Here are the transformation steps:
+* Read the image: this will return a NDArray shaped as (image height, image width, 3), with the three channels in RGB order.
+* Resize the shorter edge of the image 224.
+* Crop, using a size of 224x224 from the center of the image.
+* Shift the mean and standard deviation of our color channels to match the ones of the dataset the network has been trained on.
+* Reshape the array from (Height, Width, 3) to (3, Height, Width).
```
Comment (safrooze) [inline docs/tutorials/gluon/pretrained_models.md]: Transpose, not Reshape.

File: docs/tutorials/gluon/pretrained_models.md [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+
+```python
+def transform(image):
+    resized = mx.image.resize_short(image, 224) #minimum 224x224 images
+    cropped, crop_info = mx.image.center_crop(resized, (224, 224))
+    normalized = mx.image.color_normalize(cropped.astype(np.float32)/255,
+                                      mean=mx.nd.array([0.485, 0.456, 0.406]),
+                                      std=mx.nd.array([0.229, 0.224, 0.225])) 
+    # the network expect batches of the form (N,3,224,224)
+    flipped_axis = normalized.transpose((2,0,1))  # Flipping from (224, 224, 3) to (3, 224, 224)
```
Comment (safrooze) [inline docs/tutorials/gluon/pretrained_models.md]: Maybe a better name than "flipped_axis"? I don't really think of transpose as flipping. You're not reversing that axis to consider it a flipped axis. Maybe swapped_axis? Also perhaps fix the comment as well to match.

File: docs/tutorials/gluon/pretrained_models.md [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+    batchified = flipped_axis.expand_dims(axis=0) # change the shape from (3, 224, 224) to (1, 3, 224, 224)
+    return batchified
+```
+
+## Testing the different networks
+We run the image through each pre-trained network. The models output a *NDArray* holding 1,000 activation values, which we convert to probabilities using the `softmax()` function, corresponding to the 1,000 categories it has been trained on. The *NDArray* has only one line since batch size is equal to 1
```
Comment (safrooze) [inline docs/tutorials/gluon/pretrained_models.md]: "The *NDArray* has only one line..." to "The output prediction *NDArray* has only one row..."

File: docs/tutorials/gluon/pretrained_models.md [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+
+
+`CPU times: user 156 ms, sys: 0 ns, total: 156 ms` <!--notebook-skip-line-->
+
+`Wall time: 77.1 ms` <!--notebook-skip-line-->
+
+
+As you can see, pre-trained networks produce slightly different predictions, and have different run-time. In this case, MobileNet is almost **5 times faster** than DenseNet!
+
+## Fine-tuning using pre-trained models
```
Comment (safrooze) [inline docs/tutorials/gluon/pretrained_models.md]: Fine-tuning pre-trained models (drop using)

File: docs/tutorials/gluon/pretrained_models.md [hunk 10]
```diff
@@ -0,0 +1,373 @@
...
+import matplotlib.pyplot as plt
+import numpy as np
+import json
+%matplotlib inline
+```
+
+## Loading the model
+
+The [Gluon Model Zoo](https://mxnet.incubator.apache.org/api/python/gluon/model_zoo.html) provides a collection of off-the-shelf models. You can get the ImageNet pre-trained model by using `pretrained=True`. 
+If you want to train on your own classification problem from scratch, you can get an untrained network with a specific number of classes using the `classes` parameter: for example `classes=10`.
```
Comment (safrooze) [inline docs/tutorials/gluon/pretrained_models.md]: Maybe add a line of code to show how to get the untrained model with different number of classes? Also perhaps emphasise that one cannot change number of classes and user pretrained as the same time.
